zerver/migrations/0147_realm_emoji_drop_unique_constraint.py

```diff
@@ -0,0 +1,105 @@
+        self.new_path_template = "{realm_id}/emoji/images/{emoji_file_name}"
+
+    def copy_files(self, src_path: str, dst_path: str) -> None:
+        raise NotImplementedError()
+
+    def ensure_emoji_images(self, realm_id: int, old_filename: str, new_filename: str) -> None:
+        # Copy original image file.
+        old_file_path = self.old_orig_image_path_template.format(realm_id=realm_id,
+                                                                 emoji_file_name=old_filename)
+        new_file_path = self.new_orig_image_path_template.format(realm_id=realm_id,
+                                                                 emoji_file_name=new_filename)
+        self.copy_files(old_file_path, new_file_path)
+
+        # Copy resized image file.
+        old_file_path = self.old_path_template.format(realm_id=realm_id,
+                                                      emoji_file_name=old_filename)
+        new_file_path = self.new_path_template.format(realm_id=realm_id,
+                                                      emoji_file_name=new_filename)
+        self.copy_files(old_file_path, new_file_path)
+
+class LocalUploader(Uploader):
+    def __init__(self) -> None:
+        super().__init__()
+
+    @staticmethod
+    def mkdirs(path: str) -> None:
+        dirname = os.path.dirname(path)
+        if not os.path.isdir(dirname):
+            os.makedirs(dirname)
+
+    def copy_files(self, src_path: str, dst_path: str) -> None:
+        src_path = os.path.join(settings.LOCAL_UPLOADS_DIR, 'avatars', src_path)
+        self.mkdirs(src_path)
+        dst_path = os.path.join(settings.LOCAL_UPLOADS_DIR, 'avatars', dst_path)
+        self.mkdirs(dst_path)
+        shutil.copyfile(src_path, dst_path)
+
+class S3Uploader(Uploader):
+    def __init__(self) -> None:
+        super().__init__()
+        conn = S3Connection(settings.S3_KEY, settings.S3_SECRET_KEY)
+        bucket_name = settings.S3_AVATAR_BUCKET
+        self.bucket = conn.get_bucket(bucket_name, validate=False)
+
+    def copy_files(self, src_key: str, dst_key: str) -> None:
+        self.bucket.copy_key(src_key, self.bucket, dst_key)
+
+def get_uploader() -> Uploader:
+    if settings.LOCAL_UPLOADS_DIR is None:
+        return S3Uploader()
+    return LocalUploader()
+
+def get_emoji_file_name(emoji_file_name: str, emoji_id: int) -> str:
+    _, image_ext = os.path.splitext(emoji_file_name)
+    return ''.join((str(emoji_id), image_ext))
+
+def migrate_realm_emoji_image_files(apps: StateApps, schema_editor: DatabaseSchemaEditor) -> None:
+    RealmEmoji = apps.get_model('zerver', 'RealmEmoji')
+    uploader = get_uploader()
+    for realm_emoji in RealmEmoji.objects.all():
+        old_file_name = realm_emoji.file_name
+        realm_emoji.file_name = get_emoji_file_name(old_file_name, realm_emoji.id)
+        realm_emoji.save(update_fields=['file_name'])
+        uploader.ensure_emoji_images(realm_emoji.realm.id, old_file_name, realm_emoji.file_name)
```

Comment:
Probably should first ensure_emoji_images and then save to DB. What I mean to say is interchange order of the lines
`uploader.ensure_emoji_images(realm_emoji.realm.id, old_file_name, realm_emoji.file_name)`
and 
`realm_emoji.save(update_fields=['file_name'])`

Reply:
Since the filename depends on the RealmEmoji object's ID, which isn't allocated until we create the object, we can't do that.

Reply:
but we are in the migrations code. So realm_emoji objects already exist, right?

Reply:
Ahh, yes, that's right.

Reply:
Agree, will change it and also a migration reversal code.